noxfile.py

```diff
@@ -0,0 +1,261 @@
+    )
+
+
+@nox.session(reuse_venv=True)
+def flake8(session: Session) -> None:
+    """Run flake8 on cachi2 and tests directories"""
+    _install_requirements(session)
+    # TODO: Move flake8-pyproject to requirements-extras.txt
+    session.install("flake8-pyproject")
+    session.run(
+        "flake8",
+        "cachi2",
+        "tests",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(reuse_venv=True)
+def isort(session: Session) -> None:
+    """Run isort on cachi2 and tests directories and noxfile.py"""
+    _install_requirements(session)
+    session.run(
+        "isort",
+        "--check",
+        "--diff",
+        "--color",
+        "cachi2",
+        "tests",
+        "noxfile.py",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(reuse_venv=True)
+def mypy(session: Session) -> None:
+    """Run mypy on cachi2 and tests directories"""
+    _install_requirements(session)
+    session.run(
+        "mypy",
+        "--install-types",
+        "--non-interactive",
+        "cachi2",
+        "tests",
+        *session.posargs,
+        silent=True,
+    )
+
+
+@nox.session(name="unit-tests", reuse_venv=True, python=["3.9", "3.10", "3.11", "3.12"])
+def unit_tests(session: Session) -> None:
+    """Run unit tests and generate coverage report"""
+    _install_requirements(session)
+    session.run(
+        "pytest",
+        "--log-level=DEBUG",
+        "--cov=cachi2",
+        "--cov-config=pyproject.toml",
+        "--cov-report=term",
+        "--cov-report=html",
+        "tests/unit",
+        *session.posargs,
+    )
+
+
+@nox.session(name="coverage-html", reuse_venv=True)
+def coverage_html(session: Session) -> None:
+    """Open the coverage report in the browser"""
+    session.install("coverage")
+    session.run("coverage", "html")
+    session.run("xdg-open", "htmlcov/index.html", external=True)
+
+
+def _run_integration_tests(session: Session, env: dict[str, str]) -> None:
+    _install_requirements(session)
+    default_env = {
+        "CACHI2_TEST_NETRC_CONTENT": "machine 127.0.0.1 login cachi2-user password cachi2-pass",
+        "CACHI2_TEST_LOCAL_PYPISERVER": "true",
+        "CACHI2_TEST_LOCAL_DNF_SERVER": "true",
+    }
+    default_env.update(env)
+    session.run(
+        "pytest",
+        "--log-cli-level=WARNING",
+        "tests/integration",
+        *session.posargs,
+        env=default_env,
+    )
+
+
+@nox.session(name="integration-tests", reuse_venv=True)
+def integration_tests(session: Session) -> None:
+    """Run integration tests only for the affected code base in the current branch"""
+    _run_integration_tests(
+        session,
+        {},
+    )
+
+
+@nox.session(name="all-integration-tests", reuse_venv=True)
+def all_integration_tests(session: Session) -> None:
+    """Run all integration tests that are available"""
+    _run_integration_tests(
+        session,
+        {
+            "CACHI2_RUN_ALL_INTEGRATION_TESTS": "true",
```

Comment:
Aren't environment variables automatically passed through? I would expect us to only define the `integration_tests` target and let it be tweaked with an env variable, that's how I used to it anyway prior this work. I think documenting all the env vars we use anywhere in the project is paramount for that. Unless it doesn't work the way I imagine it does/should in which case my argument becomes invalid.

Reply:
With the growing amount of our custom env variables, this process becomes much more complicated. Recently, I wanted to generate data for integration tests and I had to define two or more mouthful env variables for that. The next day I would forget the exact names and I could start again looking for them somewhere in the code base.

I would rather have an abstraction layer as it is in this file. As I mentioned in the PR template each session's docstring is displayed in the help message. That seems enough IMO. Just keep it simple.

Reply:
Fair enough, I guess I need to think about this (after trying) harder and potentially propose a follow-up PR if anything comes out of that thought process, so you can disregard my earlier comment for the purposes of getting the PR accepted.